src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4792,16 +4797,24 @@ public override BoundNode VisitDeconstructionAssignmentOperator(BoundDeconstruct
             return null;
         }
 
-        private void VisitDeconstructionArguments(ArrayBuilder<DeconstructionVariable> variables, Conversion conversion, BoundExpression right)
+        private void VisitDeconstructionArguments(ArrayBuilder<DeconstructionVariable> variables, Conversion conversion, BoundExpression right, TypeWithState rightResult)
         {
             Debug.Assert(conversion.Kind == ConversionKind.Deconstruction);
 
             int n = variables.Count;
 
             if (!conversion.DeconstructionInfo.IsDefault)
             {
-                VisitRvalue(right);
-                var rightResult = ResultType;
+                // If we were passed an explicit right result, use that instead of visiting to figure it out
+                if (!TypeSymbol.Equals(rightResult.Type, _invalidType.Type, TypeCompareKind.ConsiderEverything2))
```

Comment:
>!TypeSymbol.Equals(rightResult.Type, _invalidType.Type, TypeCompareKind.ConsiderEverything2) [](start = 20, length = 92)

Can we use reference equality? Or perhaps declare `rightResult` as `TypeWithState? rightResult` and pass `null`?

Reply:
That's a much better idea. Will do that.